src/Authenticator/PersistenceInterface.php

```diff
@@ -41,4 +41,37 @@ public function persistIdentity(ServerRequestInterface $request, ResponseInterfa
      * @psalm-return array{request: \Psr\Http\Message\ServerRequestInterface, response: \Psr\Http\Message\ResponseInterface} Returns an array containing the request and response object
      */
     public function clearIdentity(ServerRequestInterface $request, ResponseInterface $response): array;
+
+    /**
+     * Impersonates a user
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request
+     * @param \Psr\Http\Message\ResponseInterface $response The response
+     * @param \ArrayAccess $impersonator User who impersonates
+     * @param \ArrayAccess $impersonated User impersonated
+     * @return array
+     */
+    public function impersonate(
+        ServerRequestInterface $request,
+        ResponseInterface $response,
+        \ArrayAccess $impersonator,
+        \ArrayAccess $impersonated
+    ): array;
+
+    /**
+     * Stops impersonation
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request
+     * @param \Psr\Http\Message\ResponseInterface $response The response
+     * @return array
+     */
+    public function stopImpersonating(ServerRequestInterface $request, ResponseInterface $response): array;
+
+    /**
+     * Returns true if impersonation is being done
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request
+     * @return bool
+     */
+    public function isImpersonating(ServerRequestInterface $request): bool;
```

Comment:
We will need a major release to extend the interfaces. We are likely going to have one of those soon those as there are a few other pull requests that require a major release.

Reply:
Since we are not going to implement impersonation for CookieAuthenticator, I have moved this code out to another interface ImpersonationInterface. I have added it to SessionAuthenticator only.

Reply:
I will be working on unit tests and please let me know if the target branch is ok or if we should make the MR to another branch. Thank you!

Reply:
This is fine. We can wait to do the major until this is merged. If folks urgently need changes they can use dev-master or pin the the sha they need.